dpcpp/components/reduction.dp.hpp

```diff
@@ -0,0 +1,301 @@
+    auto thread_result = zero<ValueType>();
+    for (auto i = tidx; i < size;
+         i += item_ct1.get_local_range().get(2) * item_ct1.get_group_range(2)) {
+        thread_result = reduce_op(thread_result, source[i]);
+    }
+    result[item_ct1.get_local_id(2)] = thread_result;
+
+    group::this_thread_block(item_ct1).sync();
+
+    // Stores the result of the reduction inside `result[0]`
+    reduce<sg_size>(group::this_thread_block(item_ct1), result, reduce_op);
+}
+
+
+/**
+ * @internal
+ *
+ * Computes a reduction using the add operation (+) on an array
+ * `source` of any size. Has to be called a second time on `result` to reduce
+ * an array larger than `block_size`.
+ */
+template <std::uint32_t cfg, typename ValueType>
+void reduce_add_array(
+    size_type size, const ValueType *__restrict__ source,
+    ValueType *__restrict__ result, sycl::nd_item<3> item_ct1,
+    UninitializedArray<ValueType, KCFG_1D::decode<0>(cfg)> *block_sum)
+{
+    reduce_array<KCFG_1D::decode<1>(cfg)>(
+        size, source, static_cast<ValueType *>((*block_sum)), item_ct1,
+        [](const ValueType &x, const ValueType &y) { return x + y; });
+
+    if (item_ct1.get_local_id(2) == 0) {
+        result[item_ct1.get_group(2)] = (*block_sum)[0];
+    }
+}
+
+template <std::uint32_t cfg = KCFG_1D::encode(256, 32), typename ValueType>
+void reduce_add_array(dim3 grid, dim3 block, size_t dynamic_shared_memory,
+                      sycl::queue *stream, size_type size,
+                      const ValueType *source, ValueType *result)
+{
+    stream->submit([&](sycl::handler &cgh) {
+        sycl::accessor<UninitializedArray<ValueType, KCFG_1D::decode<0>(cfg)>,
+                       0, sycl::access::mode::read_write,
+                       sycl::access::target::local>
+            block_sum_acc_ct1(cgh);
+
+        cgh.parallel_for(
+            sycl_nd_range(grid, block), [=](sycl::nd_item<3> item_ct1) {
+                reduce_add_array<cfg>(
+                    size, source, result, item_ct1,
+                    (UninitializedArray<ValueType, KCFG_1D::decode<0>(cfg)> *)
+                        block_sum_acc_ct1.get_pointer());
+            });
+    });
+}
+
+GKO_ENABLE_IMPLEMENTATION_CONFIG_SELECTION(reduce_add_array_config,
+                                           reduce_add_array);
+
+GKO_ENABLE_DEFAULT_CONFIG_CALL(reduce_add_array_call, reduce_add_array_config,
+                               kcfg_1d_list);
+
+
+/**
+ * Compute a reduction using add operation (+).
+ *
+ * @param exec  Executor associated to the array
+ * @param size  size of the array
+ * @param source  the pointer of the array
+ *
+ * @return the reduction result
+ */
+template <typename ValueType>
+ValueType reduce_add_array(std::shared_ptr<const DpcppExecutor> exec,
+                           size_type size, const ValueType *source)
+{
+    auto block_results_val = source;
+    size_type grid_dim = size;
+    auto block_results = Array<ValueType>(exec);
```

Comment:
Can't this form of initialization have overhead? Technically, it's doing a copy from one to the other which could be avoided.

Reply:
I think there is no copy? it is temporary workspace to store the block result from the reduction first step

Reply:
It's a move, not a copy, so if the initialization doesn't happen in-place (which I think it does), it would have negligible overhead.